vspace/src/main/java/edu/asu/diging/vspace/web/publicview/ExhibitionSlideController.java

```diff
@@ -79,21 +88,52 @@ public String slide(Model model, @PathVariable("slideId") String slideId, @PathV
         model.addAttribute("firstSlide", module.getStartSequence().getSlides().get(0).getId());
         String nextSlideId = "";
         String prevSlideId = "";
-        int slideIndex = 0;
-        slideIndex = sequenceSlides.indexOf(sildeManager.getSlide(slideId));
+        ISlide currentSlide = sildeManager.getSlide(slideId);
+        int slideIndex = sequenceSlides.indexOf(currentSlide);
         int slideSize = sequenceSlides.size();
         if (slideSize > slideIndex + 1) {
             nextSlideId = sequenceSlides.get(slideIndex + 1).getId();
         }
         if (slideIndex > 0) {
             prevSlideId = sequenceSlides.get(slideIndex - 1).getId();
         }
+        model.addAttribute("sequences",moduleManager.getModuleSequences(moduleId));
         model.addAttribute("sequence",sequenceExist);
         model.addAttribute("slides", sequenceSlides);
         model.addAttribute("currentSequenceId", sequenceId);
         model.addAttribute("nextSlide", nextSlideId);
         model.addAttribute("prevSlide", prevSlideId);
-        model.addAttribute("currentSlideCon", sildeManager.getSlide(slideId));
+        model.addAttribute("currentSlideCon", currentSlide);
+
+        if(currentSlide instanceof BranchingPoint) {
+            model.addAttribute("choices", ((BranchingPoint)currentSlide).getChoices());
+            if(choiceHistory.getFromSequenceSlideHistory().size()==0) {
+                choiceHistory.addToSequenceSlideHistory(sequenceId, slideId);
+            }
+        }
+        if(choice){
+            String lastElement = choiceHistory.removeLastElementFromSequenceSlideHistory();
+            if(lastElement.equals(sequenceId+","+slideId)){
+                choiceHistory.addToSequenceSlideHistory(lastElement);
+            }else {
+                choiceHistory.addToSequenceSlideHistory(lastElement);
+                choiceHistory.addToSequenceSlideHistory(sequenceId+","+slideId);
+            }
+        }
+        if(choiceHistory.getFromSequenceSlideHistory().size()>=2) {
+            model.addAttribute("showBackToPreviousChoice", true);
+            String lastElement = choiceHistory.removeLastElementFromSequenceSlideHistory();
+            String seondLastElement = choiceHistory.removeLastElementFromSequenceSlideHistory();
+            String[] elementValues = seondLastElement.split(",");
+            model.addAttribute("previousChoiceSequence", elementValues[0]);
+            model.addAttribute("previousChoiceSlide", elementValues[1]);
+            choiceHistory.addToSequenceSlideHistory(seondLastElement);
+            choiceHistory.addToSequenceSlideHistory(lastElement);
+        }
+
+        if(back) {
+            choiceHistory.removeLastElementFromSequenceSlideHistory();
+        }
```

Comment:
this all seems very complicate. wouldn't you just need this:
- in sequence controller:
   - pass parameter with branching point id that led to the sequence (if no branching point id is passed, check if sequence is start sequence, if not, find first branching point that links to the sequence)
  - push sequence/slide object onto stack
- in slide controller:
  - if there is a sequence/slide object on the stack, use that to generate links (only peek, no pop)
  - if back parameter exists and is true, pop last sequence/slide if slide is current branching point
- in branching point jsp:
  - add BP id to choice links
  - add back parameter to back to last BP button